src/Symfony/Component/Messenger/QueryBusTrait.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger;
+
+use Symfony\Component\Messenger\Exception\LogicException;
+use Symfony\Component\Messenger\Middleware\HandleMessageMiddleware;
+use Symfony\Component\Messenger\Stamp\HandledStamp;
+
+/**
+ * Leverages a message bus to expect a single, synchronous message handling and return its result.
+ *
+ * @author Maxime Steinhausser <maxime.steinhausser@gmail.com>
+ *
+ * @experimental in 4.2
+ */
+trait QueryBusTrait
+{
+    /** @var MessageBusInterface */
+    private $queryBus;
+
+    /**
+     * Dispatches the given message and returns the result from the handler returned value.
+     *
+     * @param object|Envelope $query The message or the message pre-wrapped in an envelope
+     *
+     * @return mixed The handler returned value
+     */
+    private function dispatchQuery($query)
+    {
+        if (!$this->queryBus instanceof MessageBusInterface) {
+            throw new LogicException(sprintf('You must provide a "%s" instance in the "%s::$queryBus" property, "%s" given.', MessageBusInterface::class, \get_class($this), \is_object($this->queryBus) ? \get_class($this->queryBus) : \gettype($this->queryBus)));
+        }
+
+        $envelope = $this->queryBus->dispatch($query);
+        /** @var HandledStamp[] $handledStamps */
+        $handledStamps = $envelope->all(HandledStamp::class);
+
+        if (!$handledStamps) {
+            throw new LogicException(sprintf('Query of type "%s" wasn\'t handled. Is the "%s" middleware properly registered in your bus?', \get_class($envelope->getMessage()), HandleMessageMiddleware::class));
```

Comment:
Message of type "%s" was handled zero times. Exactly one handler is expected when using a query bus. (HandleMessageMiddleware is an optional implementation)